ping.go

```diff
@@ -380,89 +380,93 @@ func (p *Pinger) SetLogger(logger Logger) {
 // done. If Count or Interval are not specified, it will run continuously until
 // it is interrupted.
 func (p *Pinger) Run() error {
-	logger := p.logger
-	if logger == nil {
-		logger = NoopLogger{}
-	}
-
-	var conn *icmp.PacketConn
+	var conn packetConn
 	var err error
 	if p.ipaddr == nil {
 		err = p.Resolve()
 	}
 	if err != nil {
 		return err
 	}
-	if p.ipv4 {
-		if conn, err = p.listen(ipv4Proto[p.protocol]); err != nil {
-			return err
-		}
-		if err = conn.IPv4PacketConn().SetControlMessage(ipv4.FlagTTL, true); runtime.GOOS != "windows" && err != nil {
-			return err
-		}
-	} else {
-		if conn, err = p.listen(ipv6Proto[p.protocol]); err != nil {
-			return err
-		}
-		if err = conn.IPv6PacketConn().SetControlMessage(ipv6.FlagHopLimit, true); runtime.GOOS != "windows" && err != nil {
-			return err
-		}
+	if conn, err = p.listen(); err != nil {
+		return err
 	}
 	defer conn.Close()
+
+	return p.run(conn)
+}
+
+func (p *Pinger) run(conn packetConn) error {
+	logger := p.logger
+	if logger == nil {
+		logger = NoopLogger{}
+	}
+
+	if err := conn.SetFlagTTL(); err != nil {
+		return err
+	}
 	defer p.finish()
 
-	var wg sync.WaitGroup
 	recv := make(chan *packet, 5)
 	defer close(recv)
-	wg.Add(1)
-	//nolint:errcheck
-	go p.recvICMP(conn, recv, &wg)
 
 	if handler := p.OnSetup; handler != nil {
 		handler()
 	}
 
-	timeout := time.NewTicker(p.Timeout)
-	interval := time.NewTicker(p.Interval)
-	defer func() {
-		p.Stop()
-		interval.Stop()
-		timeout.Stop()
-		wg.Wait()
-	}()
+	g := new(errgroup.Group)
 
-	err = p.sendICMP(conn)
-	if err != nil {
-		return err
-	}
+	g.Go(func() error {
+		defer p.Stop()
+		return p.recvICMP(conn, recv)
+	})
 
-	for {
-		select {
-		case <-p.done:
-			return nil
-		case <-timeout.C:
-			return nil
-		case r := <-recv:
-			err := p.processPacket(r)
-			if err != nil {
-				// FIXME: this logs as FATAL but continues
-				logger.Fatalf("processing received packet: %s", err)
-			}
-		case <-interval.C:
-			if p.Count > 0 && p.PacketsSent >= p.Count {
-				interval.Stop()
-				continue
+	g.Go(func() error {
```

Comment:
Is there a good way to extract out this anon func? I find these large inline funcs a bit ugly.

Reply:
OK, I gave this a go. Naming is ... hmm...